File: Sorting/MergeSort.py
#merge sort

#time complexity = O(nlogn)
#space complexity = O(n)

def mergeSort(arr):
    if len(arr) > 1:

        mid = len(arr)//2
        l1 = arr[:mid]
        l2 = arr[mid:]

        mergeSort(l1)
        mergeSort(l2)

        i = j = k = 0
        # add the number from divided array to original array
        while i < len(l1) and j < len(l2):
            if l1[i] < l2[j]:
                arr[k] = l1[i]
                i += 1
            else:
                arr[k] = l2[j]
                j += 1
            k += 1
        
        #add remaining numbers from l1 to the array
        while i < len(l1):
            arr[k] = l1[i]
            i += 1
            k += 1
        
        #add reamining numbers from l2 to the array
        while j < len(l2):
            arr[k] = l2[j]
            j += 1
            k += 1

if __name__ == '__main__':
    arr = [10, 4, 63, 90, 43, 20, 1, 9]
    print(arr)
    mergeSort(arr)
    print('array after merge sort: ', arr)

File: Sorting/mergeSort.cpp
// merge sort

// time complexity = O(nlogn)
// space complexity = O(n)

#include <iostream>
#include <vector>
using namespace std;

void mergeSort(vector<int> &arr) {
  if (arr.size() > 1) {

    int mid = arr.size() / 2;
    vector<int> left, right;

    for (int i = 0; i < mid; i++)
      left.push_back(arr[i]);

    for (int j = mid; j < arr.size(); j++)
      right.push_back(arr[j]);

    mergeSort(left);
    mergeSort(right);

    int i = 0, j = 0, k = 0;
    // add the number from divided array to original array
    while (i < left.size() && j < right.size()) {
      if (left[i] < right[j]) {
        arr[k] = left[i];
        i++;
      } else {
        arr[k] = right[j];
        j++;
      }
      k++;
    }

    // add remaining numbers
    while (i < left.size()) {
      arr[k] = left[i];
      i++;
      k++;
    }
    while (j < right.size()) {
      arr[k] = right[j];
      j++;
      k++;
    }
  }
}

int main() {
  vector<int> array = {10, 4, 63, 90, 43, 20, 1, 9};
  cout << "unsorted array: ";
  for (auto x : array)
    cout << x << " ";
  cout << "\n";
  mergeSort(array);
  cout << "sorted array: ";
  for (auto x : array)
    cout << x << " ";

  return 0;
}